Allow splitting control over modem clocks.

Frostie314159 opened the pull request:
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs developer guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

#### Description
This PR seeks to allow splitting the radio clock controller for all modems. Since I'm not sure what the best architectural choice is, due to the limited public knowledge on the clock tree of the modems, the first commit only contains a rough overview of the architecture.

Closes https://github.com/esp-rs/esp-hal/issues/1447

#### Testing

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -666,3 +670,76 @@ impl<'d> RadioClockController<'d> {
+///
+/// Using an AtomicU8 wrapped in a critical section mutex, isntead of a RefCell, prevents panic
+/// machinery from being generated.
+static PHY_CLOCK_REFS: critical_section::Mutex<AtomicU8> = critical_section::Mutex::new(AtomicU8::new(0));
```
You don't need the critical section around an atomic u8

Frostie314159 replied:
Looking at it now, that makes sense.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -666,3 +670,76 @@ impl<'d> RadioClockController<'d> {
+    critical_section::with(|cs| {
+        let inner = PHY_CLOCK_REFS.borrow(cs);
+        let new_phy_clock_ref_count = inner.load(Ordering::Relaxed);
+        if new_phy_clock_ref_count == 0 {
```
It's the old ref count 🙃

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -666,3 +670,76 @@ impl<'d> RadioClockController<'d> {
+        if new_phy_clock_ref_count == 0 {
+            clocks_ll::enable_phy(true);
+        }
+        inner.store(new_phy_clock_ref_count | (modem_bitmask as u8), Ordering::Relaxed);
```
As long as there is only one place where you can enable a particular clock at a time (i.e. it's not possible to enable wifi clocks twice), this is fine, but is that always ensured?

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -666,3 +670,76 @@ impl<'d> RadioClockController<'d> {
+// controller. If that doens't really make sense, this can be moved back.
+
+fn enable_phy_clock_internal(modem_bitmask: ModemBitmask) {
+    critical_section::with(|cs| {
```
What you want here is an atomic fetch_or

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -666,3 +670,64 @@ impl<'d> RadioClockController<'d> {
+    }
+}
+fn disable_phy_clock_internal(modem_bitmask: ModemBitmask) {
+    if PHY_CLOCK_REFS.fetch_and(!(modem_bitmask as u8), Ordering::Relaxed) == modem_bitmask as u8 {
+        clocks_ll::enable_phy(false);
```
There is a race condition here, if code enters `enable_phy` and another clock user enables its clock, the result is unpredictable.

Frostie314159 replied:
Should I use a critical section in this case?

Frostie314159 commented:
I'd see the use case for simply counting the refs, but one issue I see with it is, that unless we count the reference for each modem clock controller individually, just repeatedly calling the function to disable the PHY clock on one modem clock controller, would eventually disable the PHY clock, even if different modems still rely on it.

Frostie314159 commented:
I have implemented all modem clock controllers and adjusted `esp-ieee802154` to use the split API. I can revert that if necessary. The issue I currently see, that after the peripherals, the PHY clock state will remain enabled, if previously set, even though no peripheral technically holds a reference to it. The question is, whether it really makes sense to have one interface for all, since no single driver would conceivably require access to multiple modems. I also renamed the `reset_mac` function to `reset_wifi_mac`, since there was absolutely no documentation, which MAC it's supposed to reset.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -659,10 +678,177 @@ impl<'d> RadioClockController<'d> {
+
+#[cfg(any(bt, ieee802154, wifi))]
+/// Tracks which modems currently request an active PHY clock.
+static PHY_CLOCK_REF_COUNTERS: critical_section::Mutex<[AtomicU8; 3]> = critical_section::Mutex::new([const { AtomicU8::new(0) }; 3]);
```
Use `Cell<u8>` please. If you are in a critical section, these don't need to be atomic.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -659,10 +678,177 @@ impl<'d> RadioClockController<'d> {
+    }
+}
+#[cfg(wifi)]
+#[instability::unstable]
```
Move the unstable attribute to the methods, please.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -659,10 +678,177 @@ impl<'d> RadioClockController<'d> {
+#[cfg(bt)]
+impl Sealed for BtClockController<'_> {}
+#[cfg(bt)]
+impl ModemClockController for BtClockController<'_> {
```
```suggestion
#[instability::unstable]
impl ModemClockController for BtClockController<'_> {
```

bugadani reviewed (commented):
So, based on yesterdays comments, I think this would be better if `ModemClockController` methods returned a ClockGuard struct, that, when dropped, would decrement the relevant refcount, maybe with an unsafe disable method, but I'm unsure about that. That way the API would be pretty much impossible to misuse - users would request a clock to be active, and as long as they hold on to the returned handle, they could be sure that the clock is active.

Frostie314159 commented:
I also think something like a clock guard struct would make sense. I'll try to implement something like that.

bugadani commented on examples/src/bin/ieee802154_send_frame.rs:
```diff
@@ -26,7 +26,8 @@ fn main() -> ! {
     let delay = Delay::new();
 
-    let mut ieee802154 = Ieee802154::new(peripherals.IEEE802154, peripherals.RADIO_CLK);
+    let rcc = RadioClockController::new(peripherals.RADIO_CLK);
+    let mut ieee802154 = Ieee802154::new(peripherals.IEEE802154, rcc.split().ieee802154);
```
Just looking at this, do we even want manual access to RCC? Since you already store the clock refcounts in a global static, why do we care about the clock controllers in the first place? What if we removed `RADIO_CLK` altogether?

bugadani replied:
I'd just implement all relevant code for the radio singletons, then only radio users could meddle with any of this. What do you think about that?

Frostie314159 replied:
So you mean remove the `RadioClockController` entirely, implement `ModemClockController` for the `WIFI`, `BT` and `IEEE802154` peripheral structs and keep the PHY clock ref count in place?

Frostie314159 commented:
I've made some major adjustments to the code, namely removing the `RadioClockController` entirely, moving the functionality to the peripheral structs, making `init_radio_clocks` a public function and added a function to unsafely reduce the PHY clock ref count for a modem. I integrated these changes with esp-hal and esp-ieee802154, and updated a handful of examples.

Frostie314159 commented on esp-wifi/src/common_adapter/mod.rs:
```diff
@@ -341,10 +342,9 @@ pub(crate) unsafe fn phy_enable_clock() {
 pub(crate) unsafe fn phy_disable_clock() {
     let count = PHY_CLOCK_ENABLE_REF.fetch_sub(1, Ordering::Release);
```
Yeah right, we only need one now.

Frostie314159 replied:
I've reduced it to one.

Frostie314159 commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,164 @@ impl Clocks {
 #[cfg(any(bt, ieee802154, wifi))]
 #[instability::unstable]
-pub struct RadioClockController<'d> {
-    _rcc: crate::peripherals::RADIO_CLK<'d>,
```
I've removed it with the latest commit.

Frostie314159 commented on esp-metadata/devices/esp32.toml:
```diff
@@ -123,197 +116,345 @@ constant_0_input = 0x30
+        "rtc",
+        "touch",
+    ], alternate_functions = { 1 = "HSPIHD", 3 = "HS2_DATA1", 4 = "SD_DATA1", 5 = "EMAC_TX_ER" } },
+    { pin = 5, kind = [
+        "input",
+        "output",
+    ], alternate_functions = { 1 = "VSPICS0", 3 = "HS1_DATA6", 5 = "EMAC_RX_CLK" } },
```
Oh crap, yeah, I'll fix it. I haven't yet figured out how to disable the extension for neovim.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,135 @@ impl Clocks {
+    critical_section::with(|cs| {
+        let phy_clock_ref_counter = PHY_CLOCK_REF_COUNTERS.borrow(cs);
 
+        let new_phy_clock_ref_count = phy_clock_ref_counter.get().checked_sub(1).expect("PHY clock ref count underflowed. Either you forgot a PhyClockGuard, or used ModemClockController::decrease_phy_clock_ref_count incorrectly.");
```
Let use the `unwrap!` macro here. You can still print a message, but it'll be defmt-formatted if that's enabled.

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,135 @@ impl Clocks {
-        clocks_ll::enable_bt(enable);
+    /// Unsafely decrease the PHY clock reference count for this modem.
+    ///
+    /// # Safety
```
This function is not `unsafe`, so you don't need a `# Safety` comment. On the other hand, if this may panic, you need a `# Panics`. We also shouldn't really recomment using mem::forget here, as we explicitly say not to use it in the crate docs.

bugadani replied:
please don't, let's not encourage bad behaviour

bugadani commented on esp-wifi/src/ble/os_adapter_esp32c2.rs:
```diff
@@ -118,16 +118,16 @@ pub(super) unsafe extern "C" fn esp_intr_alloc(
 pub(super) unsafe extern "C" fn esp_reset_rpa_moudle() {
     trace!("esp_reset_rpa_moudle");
-    // stealing RADIO_CLK is safe since it is passed (as mutable reference or by
+    // stealing BT is safe since it is passed (as mutable reference or by
```
These comments no longer needs the parens, as peripherals are now always passed by value

bugadani reviewed (commented):
I'm pretty happy with how this looks, although I still have a few technicalities

Frostie314159 commented:
What should I do about the migration guide?

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
 
-/// Control the radio peripheral clocks
 #[cfg(any(bt, ieee802154, wifi))]
+/// Tracks which modems currently request an active PHY clock.
```
Technically this isn't true anymore, this just track how many users keep the PHY clock alive :)

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
+static PHY_CLOCK_REF_COUNTERS: critical_section::Mutex<Cell<u8>> =
+    critical_section::Mutex::new(Cell::new(0));
+
+// These functions are moved out of the trait to prevent monomorphization for
+// every modem clock controller. If that doens't really make sense, this can be
+// moved back.
```
This is fine as a fact, although I'm not sure how much value this comment has in the source :)

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
+        let phy_clock_ref_counter = PHY_CLOCK_REF_COUNTERS.borrow(cs);
+        let phy_clock_ref_count = phy_clock_ref_counter.get();
+
+        // If other modems have reference to the PHY clock, but this one doesn't, the
+        // PHY clock will just be set to enabled again, which shouldn't have any
+        // effect.
```
This comment also seems outdated

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
+}
+/// Do any common initial initialization needed for the radio clocks
+#[inline]
+pub fn init_radio_clocks() {
```
Please mark all new public api with `instability::unstable`. Could this maybe be called automatically, without exposing a new function? Do we need a deinit counterpart when we're tearing down esp-wifi?

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
-    #[inline]
-    pub fn enable_bt(&mut self, enable: bool) {
-        clocks_ll::enable_bt(enable);
+    /// Unsafely decrease the PHY clock reference count for this modem.
```
```suggestion
    /// Decreases the PHY clock reference count for this modem ignoring currently alive [PhyClockGuard]s.
```

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
+    /// Unsafely decrease the PHY clock reference count for this modem.
+    ///
+    /// # Panics
+    /// If the ref count is lower than the number of [PhyClockGuard]s, dropping a guard can now panic.
```
```suggestion
    /// This function panics if the PHY clock is inactive. If the ref count is lower than the number of alive [PhyClockGuard]s, dropping a guard can now panic.
```

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -591,65 +599,136 @@ impl Clocks {
-    #[inline]
-    pub fn enable_ieee802154(&mut self, enable: bool) {
-        clocks_ll::enable_ieee802154(enable);
+    /// Reset the Wi-Fi MAC
```
```suggestion
    /// Reset the Wi-Fi MAC.
```

bjoernQ commented on esp-wifi/src/common_adapter/mod.rs:
```diff
@@ -305,10 +305,15 @@ static PHY_CLOCK_ENABLE_REF: AtomicU32 = AtomicU32::new(0);
-        let radio_clocks = unsafe { RADIO_CLK::steal() };
-        RadioClockController::new(radio_clocks).enable_phy(true);
+        // stealing WIFI or BT is safe since it is passed into `init`
+        #[cfg(not(esp32h2))]
+        let peripheral = unsafe { esp_hal::peripherals::WIFI::steal() };
+        #[cfg(esp32h2)]
+        let peripheral = unsafe { esp_hal::peripherals::BT::steal() };
+
+        let clock_guard = peripheral.enable_phy_clock();
+        core::mem::forget(clock_guard);
```
I guess it's fine code wise. We should just have a comment why we use potentially unrelated peripherals (e.g. when using BT and we use WIFI here) and why that is fine

WIFI and BT are also not passed into `init` - they are passed into `wifi::new` / `BleConnector::new` (we have a few of similar comments which are now not exactly correct)

bugadani commented on esp-wifi/src/common_adapter/mod.rs:
```diff
@@ -302,13 +302,26 @@ pub unsafe extern "C" fn strrchr(_s: *const (), _c: u32) -> *const u8 {
+    #[cfg(not(esp32h2))]
+    unsafe { esp_hal::peripherals::WIFI::steal() }
+    #[cfg(esp32h2)]
+    unsafe { esp_hal::peripherals::BT::steal() }
```
Hmm apparently, maybe, the unsafe block makes this right 🤔

bjoernQ commented:
I think this looks really good - one last remaining nitpick:

There are still some comments like "stealing .. is safe since it is passed into `init`" - no peripheral is passed to `init` anymore with this PR

bjoernQ reviewed (approved):
Thanks

bugadani commented on esp-hal/src/clock/mod.rs:
```diff
@@ -602,65 +610,131 @@ impl Clocks {
-    pub fn enable_ieee802154(&mut self, enable: bool) {
-        clocks_ll::enable_ieee802154(enable);
+    /// Reset the Wi-Fi MAC.
+    pub fn reset_wifi_mac(&mut self) {
```
A bit more nit picking, sorry, but these don't need to take `&mut self` - only a single peripheral singleton should exist at a time, so every reference is technically exclusive.

Frostie314159 replied:
🥲 i'll fix it, once I'm back at my laptop

bugadani replied:
actually, let's just keep this as is, the peripehrals are `Sync` currently and accepting `&self` would actually be a problem.

bugadani commented:
Thanks a lot for this :)
